web/source/kmwembedded.ts

```diff
@@ -373,27 +374,25 @@ namespace com.keyman.text {
       // Changes for Build 353 to resolve KMEI popup key issues
       keyName=keyName.replace('popup-',''); //remove popup prefix if present (unlikely)
 
+      // Regex for 'display layer'-'virtual key name'+'optional functional layer'
       // Can't just split on '-' because some layers like ctrl-shift contain it.
-      let separatorIndex = keyName.lastIndexOf('-');
-      var displayLayer = core.keyboardProcessor.layerId;
-      if (separatorIndex > 0) {
-        displayLayer = keyName.substring(0, separatorIndex);
-        keyName = keyName.substring(separatorIndex+1);
-      }
-      if(displayLayer == 'undefined') {
-        displayLayer=core.keyboardProcessor.layerId;
+      // Virtual key starts with T_, K_, or U_
+      // matches[1]: displayLayer
+      // matches[2]: keyId
+      // matches[3]" optional functionLayer
+      let matches = keyName.match(/^(.+)-([TKU]_[^+]+)\+?(.+)?$/);
+      if (matches == null) {
+        return false
       }
```

Comment:
This is slightly different behaviour than the previous code. In the previous code, if there was no `displayLayer`, it would use `core.keyboardProcessor.layerId`, rather than bailing out. Do we need to maintain this distinction? (My preference is yes unless we can demonstrate that this never happens through further code analysis)

Reply:
I considered using a second regex if displayLayer was undefined. I just don't know what all the possible combinations of displayLayer, keyId, and functional layer there really are.

Would be something like
```
let matches = keyName.match(/^(.+)-([TKU]_[^+]+)\+?(.+)?$/);
if (matches == null) {
   matches = keyName.match(/^()([TKU]_[^+]+)\+?(.+)?$/);   // can I use an empty group so matches[1] isn't used?
   if (matches == null) {
      // both regex failed so exit
      return false
   }
}
let displayLayer = matches[1] ? matches[1]  : core.keyboardProcessor.layerId;
keyName = matches[2] + (matches[3] ? '+' + matches[3] : '');

// Determine the functional layer
let functionalLayer = (matches[3]) ? matches[3] : displayLayer;
```

      // Determine the functional layer
      let functionalLayer = (matches[3]) ? matches[3] : displayLayer;